okhttp/src/jvmTest/java/okhttp3/HappyEyeballsTest.kt

```diff
@@ -0,0 +1,242 @@
+/*
+ * Copyright (C) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3
+
+import java.io.IOException
+import java.net.Inet4Address
+import java.net.Inet6Address
+import java.net.InetAddress
+import kotlin.test.assertFailsWith
+import mockwebserver3.MockResponse
+import mockwebserver3.MockWebServer
+import org.assertj.core.api.Assertions.assertThat
+import org.junit.jupiter.api.AfterEach
+import org.junit.jupiter.api.BeforeEach
+import org.junit.jupiter.api.Test
+import org.junit.jupiter.api.extension.RegisterExtension
+import org.opentest4j.TestAbortedException
+
+/**
+ * This test binds two different web servers (IPv4 and IPv6) to the same port, but on different
+ * local IP addresses. Requests made to `127.0.0.1` will reach the IPv4 server, and requests made to
+ * `::1` will reach the IPv6 server.
+ *
+ * By orchestrating two different servers with the same port but different IP addresses, we can
+ * test what OkHttp does when both are reachable, or if only one is reachable.
+ *
+ * This test only runs on host machines that have both IPv4 and IPv6 addresses for localhost.
+ */
+class HappyEyeballsTest {
+  @RegisterExtension
+  val clientTestRule = OkHttpClientTestRule()
+
+  // Don't use JUnit 5 test rules for these; otherwise we can't bind them to a single local IP.
+  private lateinit var localhostIpv4: InetAddress
+  private lateinit var localhostIpv6: InetAddress
+  private lateinit var serverIpv4: MockWebServer
+  private lateinit var serverIpv6: MockWebServer
+
+  private val listener = RecordingEventListener()
+  private lateinit var client: OkHttpClient
+  private lateinit var url: HttpUrl
+
+  /**
+   * This is mutable and order matters. By default, it contains [IPv4, IPv6]. Tests may manipulate
+   * it to prefer IPv6.
+   */
+  private var dnsResults = listOf<InetAddress>()
```

Comment:
If we want to do this right, we may want to allow for a more specific subclass of Dns interface.

```
interface AsyncDns: Dns {
  fun lookupHosts(): List<Future<List<InetAddress>>>
} 
```

Since 

> When a client has both IPv4 and IPv6 connectivity and is trying to
   establish a connection with a named host, it needs to send out both
   AAAA and A DNS queries.  Both queries SHOULD be made as soon after
   one another as possible, with the AAAA query made first and
   immediately followed by the A query.

   Implementations SHOULD NOT wait for both families of answers to
   return before attempting connection establishment.  If one query
   fails to return or takes significantly longer to return, waiting for
   the second address family can significantly delay the connection
   establishment of the first one.  Therefore, the client SHOULD treat
   DNS resolution as asynchronous.  Note that if the platform does not
   offer an asynchronous DNS API, this behavior can be simulated by
   making two separate synchronous queries on different threads, one per
   address family.

Reply:
Likewise in Android?  https://developer.android.com/reference/android/net/DnsResolver#query(android.net.Network,%20java.lang.String,%20int,%20int,%20java.util.concurrent.Executor,%20android.os.CancellationSignal,%20android.net.DnsResolver.Callback%3C?%20super%20java.util.List%3Cjava.net.InetAddress%3E%3E)

Reply:
I think they’re independent phases. 

Look up which hosts to attempt first, then attempt to connect to 2 of them in parallel.

Reply:
Ahh now I see what you're saying, we can use new APIs to query each set independently. Yep, let's do that too.